src/broker/log.c

```diff
@@ -599,40 +532,43 @@ static void clear_request_cb (flux_t *h, flux_msg_handler_t *mh,
                               const flux_msg_t *msg, void *arg)
 {
     logbuf_t *logbuf = arg;
-    int seq;
 
-    if (flux_request_unpack (msg, NULL, "{ s:i }", "seq", &seq) < 0)
-        goto error;
-    logbuf_clear (logbuf, seq);
+    logbuf_trim (logbuf, 0);
     flux_respond (h, msg, NULL);
     return;
-error:
-    flux_respond_error (h, msg, errno, NULL);
 }
 
 static void dmesg_request_cb (flux_t *h, flux_msg_handler_t *mh,
                               const flux_msg_t *msg, void *arg)
 {
     logbuf_t *logbuf = arg;
-    const char *buf;
-    int len;
-    int seq, follow;
+    struct logbuf_entry *e;
+    int follow;
 
-    if (flux_request_unpack (msg, NULL, "{ s:i s:b }",
-                             "seq", &seq,
+    if (flux_request_unpack (msg, NULL, "{ s:b }",
                              "follow", &follow) < 0)
         goto error;
-    if (logbuf_get (logbuf, seq, &seq, &buf, &len) < 0) {
-        if (follow && errno == ENOENT) {
-            if (logbuf_sleepon (logbuf, dmesg_request_cb, h, mh, msg, arg) < 0)
-                goto error;
-            return; /* no reply */
-        }
+
+    if (!flux_msg_is_streaming (msg)) {
+        errno = EPROTO;
         goto error;
     }
-    flux_respond_pack (h, msg, "{ s:i s:s# }",
-                               "seq", seq,
-                               "buf", buf, len);
+
+    e = zlist_first (logbuf->buf);
+    while (e) {
+        if (flux_respond_pack (h, msg, "{ s:s# }",
+                               "buf", e->buf, e->len) < 0)
+            goto error;
+        e = zlist_next (logbuf->buf);
+    }
```

Comment:
One subtle change here is that by removing the single RPC based on sequence the broker is now stuck processing the entire logbuf list when responding to a dmesg request, during which time the broker is not actively processing any other events. This probably isn't really an issue at this point, but this kind of thing usually jumps out at me as having a probably of causing problems down the road.

Given that dmesg is only accessible by instance owner, and the list size is presumably bounded, maybe not something to worry about. But I do wonder if there is a way to set a maximum on the responses that are sent synchronously, and handle remaining work asynchronously so the reactor has more fairness. The implementation could be useful in other situations as well.

Reply:
That shouldn't be a problem here, since zeromq sockets provide the underlying transport.  The messages are handed off to the zeromq thread internally, and the broker thread can continue to process reactor events.

Reply:
Oh, I was talking about the time spent iterating the list, where clearly the broker can't be processing anything else since it can't re-enter its reactor until the entire list is processed. Possibly negligible in this case so I apologize for bringing it up.

Reply:
Sorry I completely misunderstood.  Ah, we could use the prep/check/idle idiom - send one response each time through the reactor loop until the client is caught up to the head of the queue.  I'm not sure that complexity is justified here but it's something to keep in mind if we need it down the line.

Reply:
It's worth noting that the original code had this in it:

```
static int logbuf_get (logbuf_t *logbuf, int seq_index, int *seq,
                       const char **buf, int *len)
{     
    struct logbuf_entry *e = zlist_first (logbuf->buf);

    while (e && e->seq <= seq_index)
        e = zlist_next (logbuf->buf);
```

so there was a `O(n)` loop going through log entries to find the correct one for the next response, for every entry that existed on the log list this was collectively `O(n^2)`   We've reduced down to a single `O(n)`iteration, so the worst "stuck processing" is hopefully about the same (obviously there's a `flux_rpc_pack()` in the middle this time).

If we were to do the prep/check/idle idiom here, we'd probably re-introduce the above looping b/c `zlist` isn't the best data structure for keeping track of where each follower is currently in the list.